waspc/cli/Command/Watch.hs

```diff
@@ -35,19 +44,45 @@ watch waspProjectDir outDir = FSN.withManager $ \mgr -> do
   chan <- newChan
   _ <- FSN.watchDirChan mgr (SP.toFilePath waspProjectDir) eventFilter chan
   _ <- FSN.watchTreeChan mgr (SP.toFilePath $ waspProjectDir </> Common.extCodeDirInWaspProjectDir) eventFilter chan
-  listenForEvents chan currentTime
+  timeOfLastEvent <- MVar.newMVar currentTime
+  listenForEvents chan currentTime timeOfLastEvent
   where
-    listenForEvents :: Chan FSN.Event -> UTCTime -> IO ()
-    listenForEvents chan lastCompileTime = do
+    oneSecond :: Int
+    oneSecond = 1000000
+
+    oneSecondDelay :: MVar.MVar UTCTime -> IO ()
+    oneSecondDelay timeOfLastEventMVar = do
+      threadDelay oneSecond
+      currentDelayTime <- MVar.readMVar timeOfLastEventMVar
+      currentTime <- getCurrentTime
+      let timeDiff = nominalDiffTimeToSeconds $ diffUTCTime currentTime currentDelayTime
+      when (timeDiff < fromIntegral oneSecond) $ do
+        threadDelay (oneSecond - (floor . (* 1e9) $ timeDiff))
```

Comment:
You should remove this line from here and put it into line 55. Right now you always get 1 second of delay + whatever you calculate here which is always too much.